worker/src/lib/util/cache/cacheFunctions.ts

```diff
@@ -132,53 +132,64 @@ export async function getCachedResponse(
   request: HeliconeProxyRequest,
   settings: { bucketSize: number },
   cacheKv: KVNamespace,
-  cacheSeed: string | null
+  cacheSeed: string | null,
 ): Promise<Response | null> {
-  const { requests: requestCaches, freeIndexes } = await getMaxCachedResponses(
-    request,
-    settings,
-    cacheKv,
-    cacheSeed
-  );
-  if (freeIndexes.length > 0) {
-    return null;
-  } else {
-    const cacheIdx = Math.floor(Math.random() * requestCaches.length);
-    const randomCache = requestCaches[cacheIdx];
-    const cachedResponseHeaders = new Headers(randomCache.headers);
-    cachedResponseHeaders.append("Helicone-Cache", "HIT");
-    cachedResponseHeaders.append(
-      "Helicone-Cache-Bucket-Idx",
-      cacheIdx.toString()
-    );
-
-    const cachedStream = new ReadableStream({
-      start(controller) {
-        let index = 0;
-        const encoder = new TextEncoder();
-
-        function pushChunk() {
-          if (index < randomCache.body.length) {
-            const chunk = encoder.encode(randomCache.body[index]);
-            controller.enqueue(chunk);
-            index++;
-            setTimeout(pushChunk, 10);
-          } else {
-            controller.close();
+  const CACHE_TIMEOUT = 2000;
+
+  try {
+    const { requests: requestCaches, freeIndexes } = await Promise.race([
+      getMaxCachedResponses(request, settings, cacheKv, cacheSeed),
+      new Promise((resolve, reject) => setTimeout(() => reject(new Error("Cache timeout")), CACHE_TIMEOUT))
+    ]) as {
+      requests: {
+        headers: Record<string, string>;
+        body: string[];
+      }[];
```

Comment:
This is fine here because the promise is returned only if we're actually returning cached data. If promise is rejected, we directly jump to the catch block, so no type mismatch.